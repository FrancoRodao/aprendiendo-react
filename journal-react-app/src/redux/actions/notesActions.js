import { db } from "../../firebase/firebase.config"
import { imageUploadToCloudinary } from "../../helpers/imageUpload"
import { loadNotes } from "../../helpers/loadNotes"
import { types } from "../types/types"

export const startAddNewNote = () => {
    return async (dispatch, getState) => {
        const userId = getState().auth.uid

        const newNote = {
            title: 'title',
            body: 'body',
            date: new Date().getTime()
        }

        const noteRef = await db.collection(`${userId}/journal/notes`).add(newNote)
        dispatch(addNewNote(noteRef.id, newNote))
        dispatch(activeNoteAction(noteRef.id, newNote))
    }
}

export const addNewNote = (noteId, note) => ({
    type: types.notesAdd,
    payload: {
        id: noteId,
        note
    }
})

export const activeNoteAction = (id, note) => ({
    type: types.notesSetActive,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (userId) => {
    return async (dispatch) => {
        const notes = await loadNotes(userId)
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})


export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth
        const noteToFireStore = {...note}
        delete noteToFireStore.id

        await db.collection(`${uid}/journal/notes`).doc(`${note.id}`).update(noteToFireStore)
        dispatch(refreshNote(note))
    }
}

export const refreshNote = (note) => ({
    type: types.notesUpdate,
    payload: {
        note
    }
})

export const startUploadPicture = (file) => {
    return async (dispatch, getState) => {
        const activeNote = getState().notes.active
        const imageUrl = await imageUploadToCloudinary(file)
        activeNote.imgUrl = imageUrl
        
        dispatch(startSaveNote(activeNote))
    }
}


export const startDeleteNote = (note) => {
    return async (dispatch, getState) => {
        const userId = getState().auth.uid

        await db.doc(`${userId}/journal/notes/${note.id}`).delete()
        dispatch(deleteNote(note))
    }
}

export const deleteNote = (note) => ({
    type: types.notesDelete,
    payload: {
        note
    }
})