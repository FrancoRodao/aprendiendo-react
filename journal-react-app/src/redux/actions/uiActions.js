import { types } from "../types/types"

export const startLoadingAction = () => (
    {
        type: types.uiStartLoading,
        payload: {
            loading: true
        }
    }
)


export const finishLoadingAction = () => (
    {
        type: types.uiFinishLoading,
        payload: {
            loading: false
        }
    }
)

export const errorAction = (state, message) => (
    {
        type: types.uiError,
        payload: {
            state,
            lastMsg: message
        }
    }
)

export const clearErrorsAction = () =>  (
    {
        type: types.uiError,
        payload: {
            state: false,
            lastMsg: ""
        }
    }
)