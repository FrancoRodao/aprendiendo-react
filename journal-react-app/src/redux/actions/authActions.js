import { types } from "../types/types"
import { firebase, googleAuthProvider } from "../../firebase/firebase.config";
import { clearErrorsAction, errorAction, finishLoadingAction, startLoadingAction } from "./uiActions";

export const startLogin = (email, password) => {
    return (dispatch) => {
        dispatch(startLoadingAction())

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(loginAction(user.uid, user.displayName))
                dispatch(clearErrorsAction())
            })
            .catch(e => {
                if (e.code === "auth/user-not-found" || "auth/wrong-password") {
                    dispatch(errorAction(true, "Wrong credentials"))
                    return
                }
                dispatch(errorAction(true, "Unexpected error"))
            })
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const startGoogleLogin = () => {
    return dispatch => {
        dispatch(startLoadingAction())

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                dispatch(loginAction(user.uid, user.displayName))
                dispatch(clearErrorsAction())
            })
            .catch(() => dispatch(errorAction(true, "Unexpected error")))
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const loginAction = (uid, displayname) => (
    {
        type: types.authLogin,
        payload: {
            uid,
            displayname
        }
    }
)

export const startRegisterUser = (name, email, password) => {
    return (dispatch, state) => {
        dispatch(startLoadingAction())
        return firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({
                    displayName: name,
                    photoURL: ''
                })
                dispatch(registerUserAction(user.uid, user.displayName))
                dispatch( clearErrorsAction() )
            })
            .catch(e => {
                if (e.code === "auth/email-already-in-use") {
                    dispatch(errorAction(true, "Email already registered"))
                    return
                }
                dispatch(errorAction(true, "Unexpected error"))
            })
            .then(() => state())
            .finally(() => dispatch(finishLoadingAction()))
        
    }
}

export const registerUserAction = (uid, displayname) => {
    return {
        type: types.register,
        payload: {
            uid,
            displayname
        }
    }
}


export const startLogout = () => {
    return async (dispatch) => {
        dispatch(startLoadingAction())

        firebase.auth().signOut()
            .then(() => dispatch(logoutUserAction()))
            .catch(() => dispatch(errorAction(true, "Unexpected error")))
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const logoutUserAction = () => ({
    type: types.authLogout,
    payload: {}
})

