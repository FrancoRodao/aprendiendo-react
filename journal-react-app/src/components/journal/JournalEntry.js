import moment from 'moment'
import React from 'react'
import { useDispatch } from 'react-redux'
import { activeNoteAction } from '../../redux/actions/notesActions'

export default function JournalEntry({ entry }) {

    const entryDate = moment(entry.date)
    const dispatch = useDispatch()

    const handleActiveEntry = () => {
        dispatch(activeNoteAction(entry.id, entry))
    }

    return (
        <div onClick={handleActiveEntry} className="journal__entry">
            {
                entry.imgUrl &&
                <div
                    className="journal__entry-picture"
                    style={{
                        backgroundSize: "cover",
                        backgroundImage: `url(${entry.imgUrl})`
                    }}
                ></div>
            }

            <div className="journal__entry-body">
                <p className="journal__entry-title">
                    {entry.title}
                </p>
                <p className="journal__entry-content">
                    {entry.body}
                </p>
            </div>

            <div className="journal__entry-date-box">
                <span>{entryDate.format("dddd")}</span>
                <h4>{entryDate.format("Do")}</h4>
            </div>

        </div>
    )
}
