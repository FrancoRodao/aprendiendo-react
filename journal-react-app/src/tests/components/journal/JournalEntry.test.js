import { mount } from "enzyme"

import { Provider } from "react-redux"
import thunk from "redux-thunk"
import createMockStore from "redux-mock-store"

import JournalEntry from "../../../components/journal/JournalEntry"

import { activeNoteAction } from "../../../redux/actions/notesActions"
jest.mock('../../../redux/actions/notesActions', () => ({
    activeNoteAction: jest.fn()
}))

const storeInitialState = {}

const middlewares = [thunk]
const mockStore = createMockStore(middlewares)
const store = mockStore(storeInitialState)
store.dispatch = jest.fn()

describe('JournalEntry tests', () => {

    const entryNote = {
        id: 7357,
        title: 'hi',
        body: 'world',
        imgUrl: 'https://someurl.com'
    }

    const wrapper = mount(
        <Provider store={store}>
            <JournalEntry entry={entryNote} />
        </Provider>
    )

    test('should match to snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })
    

    test('should active entry', () => {
        wrapper.find('.journal__entry').simulate('click')
        expect(activeNoteAction).toBeCalledWith(entryNote.id, entryNote)
    })

    
})
