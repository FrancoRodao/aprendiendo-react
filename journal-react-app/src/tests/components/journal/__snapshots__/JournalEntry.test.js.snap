// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JournalEntry tests should match to snapshot 1`] = `
<div
  className="journal__entry"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://someurl.com)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      hi
    </p>
    <p
      className="journal__entry-content"
    >
      world
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      9th
    </h4>
  </div>
</div>
`;
